src/python/pants/jvm/dependency_inference/artifact_mapper.py

```diff
@@ -205,7 +246,12 @@ def insert(
             # Default to exposing the `group` name as a package.
             packages = (f"{coord.group}.**",)
         for package in packages:
-            insert(mapping, package, addresses)
+            insert(mapping, package, addresses, False)
+
+    # Mark types that have strong first-party declarations as first-party
+    for tgt in all_jvm_type_providing_tgts:
+        for provides_types in tgt[JvmProvidesTypesField].value or []:
```

Comment:
Naming nit: Since the loop traverses a list, `provides_type` is more accurate of a name.